homeassistant/components/onkyo/media_player.py

```diff
@@ -233,14 +233,43 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
 
     _LOGGER.debug("Provisioning Onkyo AVR device at %s:%d", host, port)
 
+    active_zones = []
+    zone_discovery_completed = False
+
+    def async_onkyo_discover_zones_callback(message):
+        """Receive the power status of the available zones on the AVR."""
+        zone, command, _ = message
+        if command in ["system-power", "power"]:
+            # When we receive the status for a zone, it is available on the AVR
+            # So we create an entity for the zone and continue discovery
+            if zone in zones:
+                _LOGGER.debug("Discovered %s on %s, add to active zones", zone, name)
+                zone_entity = OnkyoAVR(avr, name, sources, zone, max_volume)
+                active_zones.append(zone_entity)
+                async_add_entities([zone_entity])
+
+            # When we receive the main status, we finish zone discovery
+            elif zone == "main":
+                _LOGGER.debug("Finished zone discovery on %s", name)
+                for zone in active_zones:
+                    zone.backfill_state()
+                return True
+        return False
+
     @callback
     def async_onkyo_update_callback(message):
         """Receive notification from transport that new data exists."""
         _LOGGER.debug("Received update callback from AVR: %s", message)
-        for zone in active_zones:
-            zone.process_update(message)
 
-    active_zones = []
+        # Define the zone_discovery_completed variable as non-local
+        # This way the variable stays consistent over multiple callbacks
+        nonlocal zone_discovery_completed
+
+        if zone_discovery_completed:
```

Comment:
It seems like we can simplify quite a lot by extracting the zone from the message and (pseudocode):
```python
if zone already found:
  zone.process_update(message)
else:
  async_onkyo_discover_zones_callback(message)
```
No need to keep track of whether discovery completed, etc...

Reply:
I prefer your approach, makes things a lot simpler!
Could you maybe check the updates I did?